### PR TITLE
Feature/custom count query

### PR DESCRIPTION
**Pull request by paulcwarren**
## Context
Count queries can be [slow](https://wiki.postgresql.org/wiki/Slow_Counting).   This impacts us because Spring Data's querydsl implementation for paged findAll methods performs a count query to calculate the total number of pages available.  This impacts the performance of the content cloud. 

This PR investigates support for paged findAll custom count queries that would give the content cloud platform better performance.

## How does it work?
Repositories that use querydsl must extend `QuerydslPredicateExecutor`.    The implementation behind this interface is `QuerydslJpaPredicateExecutor`.  Technically, this is a [fragment](https://www.baeldung.com/spring-data-composable-repositories), but it is explicitly configured during the initialization of the repository bean (rather than being detected through classpath scanning).  Therefore, we need to provide a custom RepositoryBeanFactory class `AbacJpaRepositoryFactoryBean` in order to override this underlying querydsl impl class.

Once configured in the overridden querydsl impl class `AbacQuerydslJpaPredicateExecutor` accepts an optional `CustomCountQueryProvider` bean that allows the findAll method to delegate to to get the custom count query.  I also provide a default implementation `PostgresQueryPlanEstimateCountQueryProvider` that uses this query planner count technique explained in [this article](https://www.citusdata.com/blog/2016/10/12/count-performance/#distinct_counts_exact) to execute a much faster count query. 

## How would a Content Cloud Application use this?
The application configures the custom repository factory bean:

`@EnableJpaRepositories(repositoryFactoryBeanClass = AbacJpaRepositoryFactoryBean.class)`

and if it wants to, provides a bean that implements `CustomCountQueryProvider`

## Next Steps
If we think this is useful then we can probably wrap this configuration requirement up in the spring boot starter that we want to create anyway.

We would also want to give some thought to how the count_estimate function (that backs the `PostgresQueryPlanEstimateCountQueryProvider` implementation) gets created in the database.  One option would be to use a DatabaseInitializer in the same spring boot starter that creates this function if it detects a postgres DB.  But I havent thought to hard about this yet.

Thoughts/ comments welcome.

### CODE REVIEW

**Comment by tgeens** [comment]
Can you fix the link to the referenced article ?

File: thunx-spring/src/main/java/org/springframework/data/jpa/repository/support/AbacQuerydslJpaPredicateExecutor.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.querydsl.EntityPathResolver;
+import org.springframework.data.repository.support.PageableExecutionUtils;
+import org.springframework.util.Assert;
+
+import javax.persistence.EntityManager;
+import javax.persistence.Query;
+
+public class AbacQuerydslJpaPredicateExecutor<T> extends QuerydslJpaPredicateExecutor<T> {
```
Comment (tgeens) [inline thunx-spring/src/main/java/org/springframework/data/jpa/repository/support/AbacQuerydslJpaPredicateExecutor.java:16]: Is this custom querydsl-predicate-executor still directly tied to abac/thunx ?
Or can we _promote_ this to an alternative querydsl-predicate-executor, that can be used wether we're using thunx or not ?

File: thunx-spring/src/test/java/org/springframework/data/jpa/repository/support/PostgresQueryPlanEstimateCountQueryProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+@ExtendWith(SpringExtension.class)
+@DataJpaTest
+@AutoConfigureTestDatabase(replace= AutoConfigureTestDatabase.Replace.NONE)
+@ActiveProfiles("test")
+public class PostgresQueryPlanEstimateCountQueryProviderTest {
+
+    @ClassRule
+    public static PostgreSQLContainer postgres = new PostgreSQLContainer<>("postgres");
```
Comment (tgeens) [inline thunx-spring/src/test/java/org/springframework/data/jpa/repository/support/PostgresQueryPlanEstimateCountQueryProviderTest.java:32]: Nice :+1:

File: thunx-spring/src/main/java/org/springframework/data/jpa/repository/support/AbacQuerydslJpaPredicateExecutor.java (hunk 1 above)
Comment (paulcwarren) [reply thunx-spring/src/main/java/org/springframework/data/jpa/repository/support/AbacQuerydslJpaPredicateExecutor.java:16]: No, it isn't really tied to thunx at all.  Its definitely an orthogonal concern.  If you wanted to move it out and into some other xenit repo that would work I think.

**Pull request closed without merging**